src/snowflake/connector/result_set.py

```diff
@@ -63,41 +63,27 @@ def result_set_iterator(
     """
 
     with ThreadPoolExecutor(prefetch_thread_num) as pool:
-        # Fill up window
-
         logger.debug("beginning to schedule result batch downloads")
-
         for _ in range(min(prefetch_thread_num, len(unfetched_batches))):
             logger.debug(
                 f"queuing download of result batch id: {unfetched_batches[0].id}"
             )
             unconsumed_batches.append(
                 pool.submit(unfetched_batches.popleft().create_iter, **kw)
             )
-
         yield from first_batch_iter
-
+        while unfetched_batches:
+            logger.debug(
+                f"queuing download of result batch id: {unfetched_batches[0].id}"
+            )
+            future = pool.submit(unfetched_batches.popleft().create_iter, **kw)
+            unconsumed_batches.append(future)
+        _, _ = wait(unconsumed_batches, return_when=ALL_COMPLETED)
         i = 1
         while unconsumed_batches:
-            logger.debug(f"user requesting to consume result batch {i}")
-
-            # Submit the next un-fetched batch to the pool
-            if unfetched_batches:
-                logger.debug(
-                    f"queuing download of result batch id: {unfetched_batches[0].id}"
-                )
-                future = pool.submit(unfetched_batches.popleft().create_iter, **kw)
-                unconsumed_batches.append(future)
-
-            future = unconsumed_batches.popleft()
```

Comment:
Original code wait for the left most thread to complete, thus not fully paralleled and being slow